# ch01-04. pandas

###### 목차
1. pandas로 csv파일 불러오기
2. dataframe과 리스트, 딕셔너리, 넘파이 ndarray상호 변환
3. dataframe의 칼럼 데이터 생성과 수정
4. dataframe에서의 데이터 삭제
5. index객체
6. 데이터 셀렉션 및 필터링
7. 정렬, aggregarion, groupby
8. 결손 데이터 처리
9. apply lambda 식으로 데이터 가공 및 유용한 모듈

### 1. pandas로 csv파일 불러오기, 데이터 확인

pandas 모듈 내부에서의 자료형

1. Series : DataFrame의 하나의 열만을 가져온 형태와 동일
2. DataFrame : 기본적으로 행, 열로 이루어진(엑셀과 같은) 데이터 구조

In [1]:
import pandas as pd

In [67]:
titanic_df=pd.read_csv("train_1231.csv") #csv 파일 불러오기 
titanic_df.head(10)  #10개의 데이터만 확인, default=5

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
3            4         1       1  Futrelle, M...  female  35.0      1      0          113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0          330877   8.4583   NaN        Q
6            7         0       1  McCarthy, M...    male  54.0      0      0           17463  51.8625   E46        S
7            8         0       3  Palsson, Ma...    male   2.0      3      1          349909  21.0750   NaN        S
8            9         1       3  Johnson, Mr...  female  27.0      0      2          347742  11.1333   NaN        S
9           10         1       2  Nasser, Mrs...  female  14.0      1      0          237736  30.0708   NaN        C

In [5]:
print('DataFrame의 크기:', titanic_df.shape) #행과 열의 크기 :891개의 로우와 12개의 칼럼

DataFrame의 크기: (891, 12)


In [6]:
titanic_df.info() #각 칼럼의 타입, null의 갯수, 데이터의 분포 확인

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


RangeIndex: 891 entries, 0 to 890  ***총 891개의 데이터를 가진다

Data columns (total 12 columns):  ***총 12개의 col

Age            714 non-null float64  ***714개의 non-null데이터가 있으며, type(age)=float64

In [9]:
titanic_df.describe() #기초통계량 확인, object type(문자열)은 자동으로 제외하고 나타낸다

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [10]:
#범주형 변수인지를 describe()를 통해 판단이 가능하다(기본적인 기초통계량이므로) ex.titanic_df['Pclass']
#지정된 칼럼의 데이터값 건수를 나타내는 방법 .value_counts()
#범주형 변수 칼럼에 대해 보기 쉽다
titanic_df['Pclass'].value_counts() 

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [13]:
titanic_pclass=titanic_df['Pclass'] #dataframe에서 한 열만을 나타내면 type이 Series객체를 반환한다
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [15]:
titanic_pclass.head() #인덱싱한 결과, index값+인덱싱하여 추출한 col값의 데이터

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [17]:
value_counts=titanic_df['Pclass'].value_counts() #vlaue_counts는 Series자료형에서만 지정가능
print(type(value_counts))
print(value_counts) #반환되는 형태 Series

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


&nbsp; 다른 사용자가 작성한 코드를 보다보면 언제든지 한번도 경험해보지 못한 Type을 대면할 수도 있다. 실제 커널에서 마주하게 되는 Type은 list, dictionary 같은 Python built-in type들도 있겠지만, nd.array, pd.DataFrame 같은 type을 더 많이 마주 했기 때문. 또한 세상에는 무수히 많은 Python libraies가 존재..


1. **type()** 함수를 사용하여, type의 이름을 알아본다. (구글링)
2. **dir()** 함수를 사용하여, 객체가 보유하고 있는 모든 요소를 분석한다. (문서 검색)
3. **inspect** 모듈을 사용하여 해당 객체의 구현체를 직접 분석한다. (소스 분석)

### (1)   type()
type()함수는 객체를 전달받아 해당 객체의 이름을 반환해 준다. 그리고 해당 객체 이름을 구글링.. 대부분의 객체는 Document가 있어 자세한 내용을 살펴볼 수 있음 

In [156]:
print(type(str()))

<class 'str'>


### (2)   dir() 
객체가 소유한 모든 변수와 함수리스트를 반환. 해당 객체가 어떤 정보를 보유하고 어떤 기능을 제공하는지 직접 분석한다. 해당 객체 혹은 라이브러리의 Document에 가면 각각 함수에 대한 정의 내용을 검색가능하다.

In [158]:
print(dir(str())) #객체에서의 매서드 

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


### (3) inspect 모듈
간혹 다른 사용자가 직접 정의한 모듈을 불러와 사용하는 경우가 있을 수도 있다. 이러한 경우는 커널상에 주요 내용이 많이 생략되어 이해하기 어렵다. 이러한 경우 우리는 inspect 모듈을 사용해 커널 이면에 숨겨진 코드를 분석해 볼 수 있다.

In [159]:
import inspect, collections
print(inspect.getsourcefile(collections))

C:\Users\user\.conda\envs\NLPApps\lib\collections\__init__.py


In [160]:
import inspect, collections
source =inspect.getsource(collections)
print(source)

'''This module implements specialized container datatypes providing
alternatives to Python's general purpose built-in containers, dict,
list, set, and tuple.

* namedtuple   factory function for creating tuple subclasses with named fields
* deque        list-like container with fast appends and pops on either end
* ChainMap     dict-like class for creating a single view of multiple mappings
* Counter      dict subclass for counting hashable objects
* OrderedDict  dict subclass that remembers the order entries were added
* defaultdict  dict subclass that calls a factory function to supply missing values
* UserDict     wrapper around dictionary objects for easier dict subclassing
* UserList     wrapper around list objects for easier list subclassing
* UserString   wrapper around string objects for easier string subclassing

'''

__all__ = ['deque', 'defaultdict', 'namedtuple', 'UserDict', 'UserList',
            'UserString', 'Counter', 'OrderedDict', 'ChainMap']

import _collections_abc

## 2. DataFrame과 리스트, 딕셔너리, 넘파이, ndarray 상호변환

기본적으로 dataframe은 파이썬의 리스트, 딕셔너리, numpy의 ndarray 등 다양한 데이터로부터 생성이 가능하고, 이런 형태로 변환 또한 가능하다. 여기서 dataframe과 numpy의 ndarray의 상호간 변환은 매우 빈번하게 일어난다

#### (1) list,array,dictionary를 dataframe으로 변환하는 과정

In [25]:
#넘파이 ndarray, 리스트, 딕셔너리를 dataframe으로 변환하기
#1. data : 리스트나 딕셔너리, 넘파이 ndarray값을 통해 전달받음
#2. columns : 칼럼명 리스트를 입력받아 쉽게 dataframe을 생성한다
import numpy as np
import pandas as pd


col_name1=['col1'] #칼럼명 지정하기 위한 리스트
list1=[1,2,3]      #리스트->dataframe으로 변환하기 위한 리스트 생성
array1=np.array(list1) #행렬 -> dataframe으로 변환하기 위한 리스트 생성
print("array1 shape :", array1.shape) #생성한 행렬의 형태 확인

array1 shape : (3,)


In [23]:
#리스트로 dataframe생성하는 방법
df_list1=pd.DataFrame(list1, columns=col_name1)
print("1차원 리스트로 만든 dataframe:\n",df_list1)

1차원 리스트로 만든 dataframe:
    col1
0     1
1     2
2     3


In [26]:
#numpy ndarray로 dataframe 생성하는 방법
df_array1=pd.DataFrame(array1, columns=col_name1)
print("1차원 ndarray로 만든 dataframe:\n",df_array1)

1차원 ndarray로 만든 dataframe:
    col1
0     1
1     2
2     3


앞에서는 1차원 형태의 데이터를 기반으로 dataframe을 생성했었다. 이제는 2차원 형태의 데이터를 기반으로 dataframe을 생성한다. 2행 3열 형태의 리스트와 ndarray를 기반으로 dataframe을 생성하므로 칼럼명은 3개가 필요하다

In [34]:
#칼럼명 생성
col_name2=['col1','col2','col3']

#2행, 3열의 리스트와 ndarray를 생성
list2=[[1,2,3],[11,12,13]] #리스트
array2=np.array(list2) #행렬
print(array2) #생성한 행렬의 형태 확인
print(array2.shape)

[[ 1  2  3]
 [11 12 13]]
(2, 3)


In [35]:
#리스트를 데이터프레임 형태로 변환
df_list2=pd.DataFrame(list2, columns=col_name2)
print("2차원 데이터로 만든 dataframe:\n",df_list2)

2차원 데이터로 만든 dataframe:
    col1  col2  col3
0     1     2     3
1    11    12    13


In [36]:
#ndarray를 데이터프레임 형태로 변환
df_array2=pd.DataFrame(array2, columns=col_name2)
print("2차원 ndarray로 만든 dataframe:\n",df_array2)

2차원 ndarray로 만든 dataframe:
    col1  col2  col3
0     1     2     3
1    11    12    13


In [37]:
#딕셔너리를 데이터프레임 형태로 변환
dict={'col1':[1,11],'col2':[2,22],'col3':[3,33]} #key : value값으로 이루어져 있기 때문에 key->column, value->data형태로 mapping
df_dict=pd.DataFrame(dict)
print("딕셔너리로 만든 dataframe:\n",df_dict)

딕셔너리로 만든 dataframe:
    col1  col2  col3
0     1     2     3
1    11    22    33


#### (2) dataframe을 ndarray,list,dictionary로 변환하는 과정

In [40]:
#dataframe을 ndarray로 변환하는 방법 : 머신러닝 패키지의 데이터를 입력받기 위해 빈번하게 사용되는 방법이다. 
#예. pca->clustering을 하기 위해 customer data를 군집하는 경우. 왜??
array3=df_dict.values #딕셔너리에서 value값만
print("df_dict_values type:",type(array3)) #'numpy.ndarray'
print(array3) #딕셔너리의 value값으로만 생성한 ndarray확인

df_dict_values type: <class 'numpy.ndarray'>
[[ 1  2  3]
 [11 22 33]]


In [41]:
#dataframe을 리스트로 변환
list3=df_dict.values.tolist() #데이터프레임 -> 리스트 .tolist()
print(list3)

[[1, 2, 3], [11, 22, 33]]


In [43]:
#dataframe을 딕셔너리로 변환
dict3=df_dict.to_dict() #데이터프레임 -> 딕셔너리 .to_dict('list')
print(dict3)

{'col1': {0: 1, 1: 11}, 'col2': {0: 2, 1: 22}, 'col3': {0: 3, 1: 33}}


dataframe->dictionary의 함수 .to_dict() 의 인자 : list, dict, series

In [44]:
#to_dict()
df=pd.DataFrame({'a': ['red', 'yellow', 'blue'], 'b': [0.5, 0.25, 0.125]})
df

a      b
0     red  0.500
1  yellow  0.250
2    blue  0.125

In [45]:
df.to_dict('dict') #the default: column names are keys, values are dictionaries of index, data pairs 

{'a': {0: 'red', 1: 'yellow', 2: 'blue'}, 'b': {0: 0.5, 1: 0.25, 2: 0.125}}

In [46]:
df.to_dict('list') #list : keys are column names, values are lists of column data

{'a': ['red', 'yellow', 'blue'], 'b': [0.5, 0.25, 0.125]}

In [47]:
df.to_dict('series') #series : list와 비슷하나 type이 series

{'a': 0       red
 1    yellow
 2      blue
 Name: a, dtype: object, 'b': 0    0.500
 1    0.250
 2    0.125
 Name: b, dtype: float64}

### 3. dataframe의 칼럼 데이터 세트 생성과 수정과정

In [59]:
#데이터프레임에서 새로운 열 생성
#새롭게 생성한 열에 0을 일괄적으로 대입
titanic_df['Age_0']=0 
titanic_df.head() #확인

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0  
3      0            113803  53.1000  C123        S      0  
4      0            373450   8.0500   NaN        S      0

seiries에 상숫값을 할당하면 series의 모든 데이터 세트에 상숫값이 일괄적으로 적용된다. ndarray에 상숫값을 대입할때 같은 값이 일괄적으로 더해지는 것과 같은 원리 

In [60]:
#기존 dataframe의 열에 있는 데이터를 활용하여 새로운 칼럼 생성
titanic_df['Age_by_10']=titanic_df['Age']*10
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  
0      0         A/5 21171   7.2500   NaN        S      0      220.0  
1      0          PC 17599  71.2833   C85        C      0      380.0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0  
3      0            113803  53.1000  C123        S      0      350.0  
4      0            373450   8.0500   NaN        S      0      350.0

In [61]:
#기존 데이터 칼럼 값도 일괄적으로 연산할 수있다. 
titanic_df['Age_by_10']=titanic_df['Age_by_10']+100
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  
0      0         A/5 21171   7.2500   NaN        S      0      320.0  
1      0          PC 17599  71.2833   C85        C      0      480.0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0  
3      0            113803  53.1000  C123        S      0      450.0  
4      0            373450   8.0500   NaN        S      0      450.0

#### dataframe의 데이터 삭제

**데이터프레임이름.drop("칼럼명",axis=0, index=None, columns=None, level=None, inplace=False. errors='raise')**

(axis=0  :행삭제
axis=1  :열삭제)

여기서 drop을 적용했을 때, 원래 데이터프레임에 적용되지 않는다.(drop의 인자중 inplace=False가 디폴트값이기 때문이다) 따라서 분석시 계속 사용하기 위해서는 새로운 변수로 전달하여 사용한다

In [62]:
titanic_drop_df=titanic_df.drop('Age_0',axis=1)
titanic_drop_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  
0      0         A/5 21171   7.2500   NaN        S      320.0  
1      0          PC 17599  71.2833   C85        C      480.0  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0  
3      0            113803  53.1000  C123        S      450.0  
4      0            373450   8.0500   NaN        S      450.0

In [63]:
#앞에서 생성한 열을 모두 삭제한다. 
drop_result=titanic_df.drop(['Age_0','Age_by_10'],axis=1, inplace=True) #inplace=True로 지정하는 경우,none값이 반환되고, 
                                                                         #원래 데이터프레임에 변경값이 적용된다
print('inplace=TRUE로 drop한 후 반환된 값:',drop_result)
titanic_df.head(3)

inplace=TRUE로 drop한 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [68]:
#행을 삭제 (인덱스 값 0,1,2를 삭제하는 예제)
pd.set_option('display.width',1000)          #option
pd.set_option('display.max_colwidth',15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))                     #행을 삭제하기 이전

titanic_df.drop([0,1,2],axis=0, inplace=True) #행의 일부를 삭제
print("#### after axis 0 drop ####")
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


### 5. Index 객체

In [71]:
#원본 파일 다시 로딩
titanic_df=pd.read_csv('train_1231.csv')

#index 객체만 추출
indexes=titanic_df.index
print(indexes)

#index객체를 실제 값 array로 변환
print("index 객체 array값:\n", indexes.values)

RangeIndex(start=0, stop=891, step=1)
index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232

In [72]:
#index객체는 식별성 데이터로 1차원의 array를 가진다. 또 ndarray와 유사하게 단일값 반환, 슬라이싱이 가능하다
print(type(indexes.values))
print(indexes.values.shape)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
6


In [73]:
indexes[0]=5 #error : 인덱싱 조작은 불가

TypeError: Index does not support mutable operations

In [74]:
#series객체 출력시 인덱스를 함께 반환함 but 연산시 index는 제외된다
#인덱스는 오직 '식별값'으로만 사용된다
series_fair=titanic_df['Fare'] #series데이터 생성
print("Fair Series max:",series_fair.max())
print("Fair Series sum:",series_fair.sum())
print("sum(Fair Series) :",sum(series_fair))
print("Fair Series +3:\n",(series_fair+3).head(3))

Fair Series max: 512.3292
Fair Series sum: 28693.9493
sum(Fair Series) : 28693.949299999967
Fair Series +3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


In [76]:
#reset_index()를 지정하면 새롭게 인덱스를 지정하고, 기존의 인덱스 값은 index 칼럼으로 새롭게 추가 된다
#인덱스 값이 연속적인 int숫자형 데이터가 아닌 경우에 새로 int형 숫자형 데이터로 만들 때 사용한다
titanic_reset_df=titanic_df.reset_index(inplace=False)
titanic_reset_df.head()

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
3      3            4         1       1  Futrelle, M...  female  35.0      1      0          113803  53.1000  C123        S
4      4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450   8.0500   NaN        S

In [77]:
#reset_index의 사용
print("#### before reset_index ####")
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)

print('value_counts 객체 변수 타입:',type(value_counts))
new_value_counts=value_counts.reset_index(inplace=False) #새로운 인덱스 생성, 기존의 value_counts의 인덱스는 새로운 열
print("### After reset_index ###")
print(new_value_counts)
print("new_value_counts객체 변수 타입:",type(new_value_counts))

#### before reset_index ####
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


### 6. 데이터 셀렉션 및 필터링

numpy의 경우 '[ ]'연산자로 단일값 추출, 슬라이싱, 팬시 인덱싱, 불린인덱싱을 수행함. 판다스의 경우는 ix[ ], iloc[ ], loc[ ]연산을 통해 동일한 작업을 수행한다.

(1) dataframe에서의 [ ]연산자 : 

   1.***칼럼지정 연산자***로 사용한다

   2.***슬라이싱***이나 ***불린인덱싱***으로 사용가능하다

In [79]:
print('단일 칼럼 데이터 추출:\n',titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n',titanic_df[['Survived','Pclass']].head(3))
#print("[]안의 숫자 index는 keyerror오류 발생한다. ",titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


In [80]:
#dataframe의 슬라이싱 : 행을 슬라이싱 한다
titanic_df[0:2]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

In [81]:
#dataframe의 불린 인덱싱
#titanic_df['Pclass']==3 값이 True, False로 반환되기 때문에 boolen indexing
titanic_df[ titanic_df['Pclass']==3 ].head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

(2) dataframe.ix[], dataframe.iloc(), dataframe.loc()

- dataframe.ix[인덱스값, 칼럼명 or 칼럼위치] : DeprecationWarning, 자주 사용하지 않는다

- dataframe.loc[] : 컬럼 **명칭**을 기반

- dataframe.iloc[] : 컬럼 **위치**를 기반

In [89]:
#DeprecationWarning경고 무시하기
#.ix를 사용하면 명칭기반 인덱싱과 위치기반 인덱싱이 헷갈리기 때문에 warning이 출력된다. 따라서 잘 안쓰는 경우이다.
#여기선 경고를 출력하지 않기위해....
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning) 

In [111]:
print("컬럼 위치 기반 인덱싱 데이터 추출 : ",titanic_df.ix[0,2])
print("컬럼 명칭 기반 인덱싱 데이터 추출 :",titanic_df.ix[0,'Pclass'])

컬럼 위치 기반 인덱싱 데이터 추출 :  3
컬럼 명칭 기반 인덱싱 데이터 추출 : 3


In [86]:
#.ix[]연산자는 ndarray의 [ ]연산자와 동일하게 단일 지정, 슬라이싱, 불린 인덱싱이 모두 가능하다
data={'Name':['Chulmin','Eunkyung','Jinwoong','Soobeom'],
     'Year':[2011, 2016, 2015, 2015],
     'Gender':['M','F','M','M']} #dict
data_df=pd.DataFrame(data, index=['one','two','three','four'])#딕셔너리->데이터프레임 전환, 인덱스 값은 문자열로 지정
data_df

Name  Year Gender
one     Chulmin  2011      M
two    Eunkyung  2016      F
three  Jinwoong  2015      M
four    Soobeom  2015      M

In [90]:
#ix[]연산 유형 적용
data_df.ix[0,0] #데이터프레임 (0,0)의 값 반환

'Chulmin'

In [91]:
data_df.ix['one',0] #행번호 대신 인덱스값 지정 가능

'Chulmin'

In [92]:
data_df.ix[3,'Name'] #4번째 행의 'Name'컬럼에서의 값

'Soobeom'

In [93]:
data_df.ix[0:2, [0,1]] #데이터프레임에서 (0:2, 0:2)의 값을 슬라이싱

Name  Year
one   Chulmin  2011
two  Eunkyung  2016

In [94]:
data_df.ix[0:3, ['Year','Name']] #컬럼명으로도 슬라이싱 가능

Year      Name
one    2011   Chulmin
two    2016  Eunkyung
three  2015  Jinwoong

In [95]:
data_df.ix[:] #전체 dataframe반환, data_df[:,:]도 동일결과

Name  Year Gender
one     Chulmin  2011      M
two    Eunkyung  2016      F
three  Jinwoong  2015      M
four    Soobeom  2015      M

In [96]:
data_df.ix[data_df.Year>=2014] #특정열을 기준으로 조건에 맞는 값만 나타내는 경우, data_df.Year,data_df['Year']은 같은 표현

Name  Year Gender
two    Eunkyung  2016      F
three  Jinwoong  2015      M
four    Soobeom  2015      M

In [97]:
data_df[data_df.Year>=2014]  #data_df.ix로 지정한 값과 동일한 결과

Name  Year Gender
two    Eunkyung  2016      F
three  Jinwoong  2015      M
four    Soobeom  2015      M

명칭기반 인덱싱과 위치기반 인덱싱의 구분
1. 명칭기반 인덱싱 : 칼럼의 명칭을 기반으로 위치를 지정하는 경우
2. 위치가반 인덱싱 : 0을 출발점으로 가로축, 세로축 좌표 기반의 행과 열 위치를 기반으로 데이터를 지정한다

In [100]:
#기존의 인덱스는 문자열이므로, int형, 숫자형으로 정렬된 새로운 index를 생성한다
data_df_reset=data_df.reset_index()
data_df_reset=data_df_reset.rename(columns={'index':'old_index'})

#인덱스 값에 1을 더하여 1부터 시작하는 새로운 인덱스를 생성
data_df_reset.index=data_df_reset.index+1
data_df_reset

old_index      Name  Year Gender
1       one   Chulmin  2011      M
2       two  Eunkyung  2016      F
3     three  Jinwoong  2015      M
4      four   Soobeom  2015      M

In [103]:
data_df_reset.ix[1,1] #'명칭기반'이므로, data_df_reset.ix[0,1]은 오류를 반환한다

'Chulmin'

ix[]의 경우는 행과 열 위치에 명칭 인덱싱과 위치기반 인덱싱을 모두 허용한다. 따라서 행 위치의 인덱싱이 integer값의 경우, 혼란스러울 수 있으니, iloc[]를 사용한다

#### DataFrame.iloc() 의 사용

DataFrame.iloc()는 **위치기반의 인덱싱**만 지원하므로 integer형태의 인자만 받는다.

In [106]:
data_df #데이터프레임의 형태 확인

Name  Year Gender
one     Chulmin  2011      M
two    Eunkyung  2016      F
three  Jinwoong  2015      M
four    Soobeom  2015      M

In [107]:
data_df.iloc[0,0] #(0,0)의 위치의 값을 반환

'Chulmin'

#### DataFrame.loc[]의 사용

DataFrame.loc[]는 **명칭기반의 인덱싱**을 지원한다.

- 행위치 : Dataframe의 인덱스 값

- 열위치 : Dataframe의 칼럼명

In [108]:
data_df.loc['one','Name']

'Chulmin'

In [109]:
data_df_reset.loc[1,'Name']

'Chulmin'

대부분 슬라이싱 기호를 사용할 때, 0:3이면 (0,1,2)를 의미한다. 즉 0:n시 n-1값까지를 나타낸다. 하지만 명칭기반 인덱싱시, (종료값-1)이 적용되지 않는다. 명칭은 숫자형이 아닐 수도 있기 때문이다.

In [116]:
print('명칭기반 ix slicing\n',data_df.ix['one':'two','Name'],'\n')
print('위치기반 iloc slicing\n',data_df.iloc[0:1,0],'\n')
print('명칭기반 loc slicing\n',data_df.loc['one':'two','Name'])

명칭기반 ix slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object 

위치기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


불린 인덱싱

ix,iloc, loc와 같이 명확히 인덱싱을 지정하는 방법보다 불린 인덱싱에 의존하야 데이터를 가져오는 경우가 더 많다. 왜냐하면 명칭이나 위치 지정 인덱싱에서 가져올 값은 주로 로직이나 조건에 의해 계산된 뒤에 ix[]내에 입력하면 자동으로 원하는 값을 필터링하기 때문이다. 그리고 불린 인덱싱을 [],ix[], loc[]에서 공통으로 지원한다.

In [119]:
titanic_df=pd.read_csv('train_1231.csv') #데이터 다시 불러오기
#titanic_df['Age']>60 는 True,False로 나타내므로 titanic_df에서 해당 조건에 맞는 경우를 한번더 인덱싱 해주어야 한다
titanic_boolean=titanic_df[titanic_df['Age']>60]
print(titanic_boolean)
titanic_boolean

     PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280     

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

In [121]:
titanic_df[titanic_df['Age']>60][['Name','Age']].head(3) #특정 열 추출시 이중 인덱싱 주의

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [123]:
titanic_df.loc[titanic_df['Age']>60, ['Name','Age']].head(3) #명칭기반 인덱싱 .loc로도 불린 인덱싱 가능

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [127]:
#복잡한 조건에도 가능하다는 예
titanic_df[ (titanic_df['Age']>60) & (titanic_df['Pclass']==1) & (titanic_df['Sex']=='female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

### 7. 정렬, Aggregation함수, GroupBy 적용

정렬

***인덱스로 정렬할 때***<br> 
obj.sort_index( )    : Series, DataFrame 모두<br> 
obj.sort_index(ascending=False)<br> 

***컬럼을 정렬할 때***<br> 
frame.sort_index(axis=1)<br>

***값으로 정렬할 때***<br> 
series.sort_value( )<br> 
frame.sort_value(by=‘col’)<br> 
frame.sort_value(by=[‘a’,’b’] )<br> 

***등수***<br> 
obj.rank( )      <br> 

(1) Dataframe, Series의 정렬 - sort_values()

In [129]:
titanic_sorted=titanic_df.sort_values(by=['Name']) #name칼럼으로 오름차순 정렬하여 반환
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

In [130]:
#여러개의 칼럼으로 정렬하는 경우
titanic_sorted=titanic_df.sort_values(by=['Name','Pclass'], ascending=False)
titanic_sorted.head()

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch         Ticket     Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0         345777   9.5000   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2       A/5. 851  14.5000   NaN        S
361          362         0       2  del Carlo, ...  male  29.0      1      0  SC/PARIS 2167  27.7208   NaN        C
282          283         0       3  de Pelsmaek...  male  16.0      0      0         345778   9.5000   NaN        S
286          287         1       3  de Mulder, ...  male  30.0      0      0         345774   9.5000   NaN        S

Aggregation 함수의 적용 : min(), max(), sum(), count()와 같은 함수를 적용하는 경우를 말한다

count <br>
min, max, median, mean, var, std <br>
cumsum <br>
diff <br>


In [131]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [133]:
titanic_df[['Age','Fare']].mean() #특정 칼럼에 Aggregation함수를 적용하는 경우

Age     29.699118
Fare    32.204208
dtype: float64

groupby()의 적용 : 특정 칼럼을 기준으로 해당 group으로 묶어 나타내게 해준다

In [134]:
titanic_groupby=titanic_df.groupby(by='Pclass').count() #'pclass'칼럼에서의 고유의 category에 해당하는 데이터의 수를 count하게 해준다
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

In [137]:
#groupby한 결과를 특정 열에 대해 필터링하는 하여 보이는 경우
titanic_groupby=titanic_df.groupby('Pclass')[['PassengerId','Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [138]:
#여러개의 aggregation함수를 적용할 경우에 agg()내에 인자로 입력해서 사용한다
titanic_df.groupby('Pclass')['Age'].agg([min,max])

min   max
Pclass            
1       0.92  80.0
2       0.67  70.0
3       0.42  74.0

In [139]:
#여러개의 칼럼에 서로다른 aggregation함수를 groupby에서 호출하는 경우, 딕셔너리로 지정하여 agg()의 인자로 지정한다
agg_format={'Age':'max','SibSp':'sum','Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

### 8. 결손 데이터 처리하기 : NULL에 대한 처리

결손 데이터는 결측값을 의미하며 NULL인 경우를 의미한다. numpy에서는 NaN으로 처리한다. 기본적으로 머신러닝 알고리즘은 이 NaN 값을 처리하지 않으므로 이 값을 다른 값으로 대체해야 한다

is.na() : 결손 데이터여부 확인

In [140]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

In [141]:
#isna().sum()이면 결손 데이터의 개수를 구할 수 있다. 
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [142]:
titanic_df.info() #그냥 이렇게 보는 것이 간편

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


fillna()로 결손 데이터 대체하기

In [144]:
titanic_df['Cabin']=titanic_df['Cabin'].fillna('C000') #titanic_df['Cabin']의 결측값을 C000으로 대체
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

In [145]:
#Age 칼럼의 NaN값을 평균 나이로, 'Embarked'칼럼의 NaN값을 S로 대체
titanic_df['Age']=titanic_df['Age'].fillna(titanic_df['Age'].mean()) #결측값을 평균으로 대체
titanic_df['Embarked']=titanic_df['Embarked'].fillna('S') #결측값을 s로 대체
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

### 9. apply lamda 식으로 데이터 가공

Lambda 함수는 익명 함수(Anonymous functions)라고도 하는데, 사실 프로그래밍 하는데 있어 꼭 알아야 하는 개념은 아니다. 하지만 종종 Lambda가 Kaggle 커널에서 심심치 않게 등장한다.

Lambda 함수에 대해서는 프로그래밍 언어학적으로 대학에서 한달 이상 시간을 할애하며 배울만큼 다양한 개념이 존재하지만, 실 사용법은 매우 단순하기 때문에 한줄로만 정의가능하다

**Lambda는 함수를 한줄로 선언할 수 있는것**
```python
number_add = lambda x1, x2 : x1 + x2
```

```python
def get_square(a):
    return a**2

print("3의 제곱은:",get_square(3))
```

In [146]:
#lambda의 사용
lambda_square=lambda x:x**2 #입력인자 x : 입력인자를 기반으로한 계산식으로 결과가 반환된다
print("3의 제곰은:",lambda_square(3))

3의 제곰은: 9


Lambda 함수를 사용하면, 코드가 간결해지고 코드의 퍼포먼스가 올라가는 장점이 존재. 그렇다고 해서 무조건 Lambda를 사용해야만 하는건 아님. 두줄 이상의 코드를 람다함수로 선언하지 못하기 때문입니다. 

### unpacking

결국 unpacking은 여러가지의 값을 가진 하나의 변수를 여러변수로 나누는 것을 의미한다.
```Python 
from sklearn import svm

params = {
    'C': 1.0, 
    'kernel': 'precomputed', 
    'gamma' : 'auto_deprecated'
}

clf = svm.SVC(**params)
```

상기 코드는 'params' 라는 딕셔너리에 SVC 생성자 함수의 인자를 정의하고, **unpacking**연산자를 사용해 딕셔너리를 SVC 생성자 함수의 파라미터에 인자로 전달하는 코드.<br> 

**unpacking**연산자는 딕셔너리의 key는 파라미터 변수 이름, value는 인자로 치환하여 정확히는 **unpacking**하여 함수를 호출해 주는 연산자입니다. 

아래 코드는 위 params 딕셔너리가 unpacking되어 실제 실행되는 코드의 예시이다.
```Python 
from sklearn import svm
clf = svm.SVC(C=1.0, kernel='precomputed', gamma='auto_deprecated')
```

In [152]:
params = (0,1,2,3,4)
def tuple_unpacking(first, second, *args): 
    print(first, second)
    print(args)
    
tuple_unpacking(*params)

0 1
(2, 3, 4)


In [161]:
params = {'name':'hyunhee', 'gender':'female', 'age':10, 'isHansome':False}

def dict_unpacking(name, gender,*argv,  **kwargs):#argv는 튜플, kwargs는 키워드(딕셔너리), positional은 위치 선정 중요
    print(name, 'is',gender)
    print(kwargs)
    
dict_unpacking(**params)

hyunhee is female
{'age': 10, 'isHansome': False}


**unpacking** 연산자를 사용하여 함수를 호출할 때의 장점은 코드가 가독성이 높아진다. 기본적으로 sklean, tensorflow 머신러닝 프레임워크같은 라이브러리들은 10개가 넘어가는 매개변수를 수정해 줘야하는 경우가 존재.. 그렇기 때문에 매개변수를 딕셔너리 리스트로 관리한다면 코드의 가독성이 올라갈 뿐 아니라 함수 호출하는 행위를 데이터로 관리 가능.

### 내장함수

Lambda가 언급된 이유는 함수형 프로그래밍 패러다임을 조금 맛을 보기 위해서 이다. 함수형 프로그래밍 패러다임이란 프로그래밍 언어에서 제공하는 내장함수를 이용해 프로그래밍 하는 것.

Python 내장함수 목록[https://docs.python.org/ko/3/library/functions.html]

케글에서 가장 많이 사용되는 map과 filter함수

### map(f, iterable)
map(f, iterable)은 함수(f)와 반복 가능한(iterable) 자료형을 인자로 받습니다. map은 입력받은 자료형의 각 요소를 함수 f가 수행한 결과를 돌려주는 함수입니다.

In [164]:
result = map(lambda x: 2*x, [1,2,3,4,5]) #인자를 함수로 전달
list(result)

[2, 4, 6, 8, 10]

map함수와 + unpacking연산자의 사용

In [165]:
from random import sample

def sentence(s, v, o):
    return f'{s} {v} {o}'

pick = lambda list_ : sample(list_, 1).pop() #랜덤추출 1개

list_s = ['he', 'she', 'i', 'you', 'we']
list_v = ['practive', 'enjoy', 'love', 'like']
list_o = ['python', 'java', 'machine learning', 'deep learning']

params =[ 
    (pick(list_s),pick(list_v), pick(list_o))
    for _ in range(10) ]

list(map(lambda argv:sentence(*argv) , params)) 
#여러개의 parameter을 선언하기 위함.(map은 하나의 인자만 받음)

['she enjoy machine learning',
 'you practive python',
 'we like java',
 'i love deep learning',
 'he love machine learning',
 'he like machine learning',
 'you enjoy deep learning',
 'i practive python',
 'i love java',
 'she like machine learning']

#### filter(f_condition, iterable )

filter란 무엇인가를 걸러낸다는 뜻으로 filter 함수도 동일한 의미!

filter 함수는 첫 번째 인수로 함수 이름을, 두 번째 인수로 그 함수에 차례로 들어갈 반복 가능한 자료형을 받는다. 그리고 두 번째 인수인 반복 가능한 자료형 요소가 첫 번째 인수인 함수에 입력되었을 때 반환 값이 참인 것만 걸러내 돌려준다

In [155]:
list(filter(lambda x: 3>x, [1,2,3,4,5]))

[1, 2]

## 마지막. 시각화 matplotlib
https://matplotlib.org/
